### PR TITLE
implement json spell checker

### LINKED ISSUE #429: Generalizing Typo Detection
- First, we should check if there is an already existing library
- If we need to develop it ourselves, it should be a separate package that can be used by any JSON Schema tooling. Give a JSON Schema, give JSON, find typos in JSON for terms that are defined in JSON Schema. The report format should be thought carefully since it is needed as input for different editors to point to the specific location. JSON Path is probably enough?
- Checking if rules on top of [spectral](https://github.com/stoplightio/spectral) can do it. We are planning to use it for linting anyways.

### PR DESCRIPTION
**Pull request by hasanheroglu**
Closes #429

### CODE REVIEW

File: packages/json-spell-checker/index.js [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+ */
+function configure(schema = SCHEMA, similarityThreshold = SIMILARITY_THRESHOLD, maxLengthDifference = MAX_LENGTH_DIFFERENCE) {
+    SCHEMA = schema
+    TYPO_LOOKUP_TABLE = createSchemaLookupTable(SCHEMA)
+    SIMILARITY_THRESHOLD = similarityThreshold
+    MAX_LENGTH_DIFFERENCE = maxLengthDifference
+}
+
+/**
+ * Checks possible typos in a TD
```
Comment (egekorkan) [inline packages/json-spell-checker/index.js:33]: the documentation here should be changed to any JSON from TD

File: packages/json-spell-checker/index.js [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+ * @returns List of possible typos where the typo consists of string value of typo itself and the message,
+ * another string value, to be prompted to the user for the fix
+ */
+function checkTypos(td) {
+    const typos = []
+
+    const lookupTable = TYPO_LOOKUP_TABLE
+    const searchDepth = 1
+    const searchPath = PATH
+    let tdJson = {}
```
Comment (egekorkan) [inline packages/json-spell-checker/index.js:44]: if code uses td, it should also be just json

File: packages/json-spell-checker/index.js [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    if (schema.type === 'array') {
+        const items = schema.items
+
+        for (const item in items) {
+            if (items.hasOwnProperty(item)) {
+                if (item === REF) {
+                    if (path[0] === 'r' && references.includes(items[item])) {
+                        continue
+                    }
+
```
Comment (egekorkan) [inline packages/json-spell-checker/index.js:245]: please add comments on why r is added

**Review by egekorkan — CHANGES_REQUESTED** [review]

File: packages/json-spell-checker/examples/open-api/open-api-example.json [hunk 4]
```diff
@@ -0,0 +1,126 @@
+{
+    "swagger": "2.0",
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply packages/json-spell-checker/examples/open-api/open-api-example.json]: Fixed at [1c053e6](https://github.com/thingweb/thingweb-playground/pull/443/commits/1c053e644fdb2da7dbe78af491261b5c37852282)

File: packages/json-spell-checker/index.js (hunk 3 above)
Comment (hasanheroglu) [reply packages/json-spell-checker/index.js:248]: Fixed at [a9f2bdd](https://github.com/thingweb/thingweb-playground/pull/443/commits/a9f2bdd96a2a0fc3404919c280cec59291109d2c)

File: packages/json-spell-checker/index.js (hunk 2 above)
Comment (hasanheroglu) [reply packages/json-spell-checker/index.js]: [a9f2bdd](https://github.com/thingweb/thingweb-playground/pull/443/commits/a9f2bdd96a2a0fc3404919c280cec59291109d2c)

File: packages/json-spell-checker/index.js (hunk 1 above)
Comment (hasanheroglu) [reply packages/json-spell-checker/index.js]: [a9f2bdd](https://github.com/thingweb/thingweb-playground/pull/443/commits/a9f2bdd96a2a0fc3404919c280cec59291109d2c)

File: packages/json-spell-checker/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  const jsonSpellChecker = require("@thing-description-playground/json-spell-checker")
+  ```
+
+  * Browser: Import the `JsonSpellChecker` as a global by adding a script tag to your HTML.
+
+  ```html
+  <script src="./node_modules/@thing-description-playground/json-spell-checker/dist/web-bundle.min.js"></script>
+  ```
+
+* First configure your spell checker using (similarityThreshold should be between 0 and 1 and try to set maxLengthDifference low to use the tool faster):
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply packages/json-spell-checker/README.md]: Updated at [12f6517](https://github.com/thingweb/thingweb-playground/pull/443/commits/12f65179b16a8fa906fb3a2be5149081635ce348)

File: packages/web/package.json [hunk 6]
```diff
@@ -26,6 +26,7 @@
     "@thing-description-playground/defaults": "^1.2.1",
     "@thing-description-playground/td_to_asyncapi": "^1.0.2",
     "@thing-description-playground/td_to_openapi": "^1.2.2",
+    "@thing-description-playground/spell_checker": "file:../json-spell-checker",
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply packages/web/package.json:29]: How are we going to use the spell checker on web?

[the remark this answers is not in the document]
Comment (hasanheroglu) [reply packages/web/package.json]: Fixed at [3c09b46](https://github.com/thingweb/thingweb-playground/pull/443/commits/3c09b46b3f065ea264a7fe4b44b0c7fac323bb71)

**Review by egekorkan — APPROVED** [review]
Tested locally and it works nicely. I will proceed with publishing all these as updates to npm

**Pull request merged**